anxcloud/resource_kubernetes_cluster.go

```diff
@@ -0,0 +1,105 @@
+package anxcloud
+
+import (
+	"context"
+	"errors"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	corev1 "go.anx.io/go-anxcloud/pkg/apis/core/v1"
+	kubernetesv1 "go.anx.io/go-anxcloud/pkg/apis/kubernetes/v1"
+	"go.anx.io/go-anxcloud/pkg/utils/pointer"
+)
+
+func resourceKubernetesCluster() *schema.Resource {
+	return &schema.Resource{
+		Description: strings.TrimSpace(`
+			Resource to create Kubernetes clusters.
+			Updates are currently not supported.
+		`),
+
+		CreateContext: tagsMiddlewareCreate(resourceKubernetesClusterCreate),
+		ReadContext:   tagsMiddlewareRead(resourceKubernetesClusterRead),
+		DeleteContext: resourceKubernetesClusterDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(45 * time.Minute),
+			Read:   schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+		Schema: withTagsAttribute(schemaKubernetesCluster()),
+	}
+}
+
+func resourceKubernetesClusterCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	a := apiFromProviderConfig(m)
+
+	cluster := kubernetesv1.Cluster{
+		Name:              d.Get("name").(string),
+		Version:           d.Get("version").(string),
+		Location:          corev1.Location{Identifier: d.Get("location").(string)},
+		NeedsServiceVMs:   pointer.Bool(d.Get("needs_service_vms").(bool)),
+		EnableNATGateways: pointer.Bool(d.Get("enable_nat_gateways").(bool)),
+		EnableLBaaS:       pointer.Bool(d.Get("enable_lbaas").(bool)),
+	}
+
+	if err := a.Create(ctx, &cluster); err != nil {
+		return diag.Errorf("failed to create Kubernetes cluster: %s", err)
+	}
+
+	d.SetId(cluster.Identifier)
+
+	if err := cluster.AwaitCompletion(ctx, a); err != nil {
+		return diag.Errorf("failed awaiting Kubernetes cluster completion: %s", err)
+	}
+
+	return resourceKubernetesClusterRead(ctx, d, m)
+}
+
+func resourceKubernetesClusterRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
```

Comment:
should we check if the cluster is progressing in some way and reflect that in terraform state?

Reply:
AwaitCompletion has been moved to `Read` (which is also called by `Create`).